probe-rs/src/debug/mod.rs

```diff
@@ -222,340 +221,498 @@ pub struct SourceLocation {
+    /// [DWARF](https://dwarfstd.org) 6.4.4 - CIE defines the return register address used in the `gimli::RegisterRule` tables for unwind operations. Theoretically, if we encounter a function that has `Undefined` `gimli::RegisterRule` for the return register address, it means we have reached the bottom of the stack OR the function is a 'no return' type of function. I have found actual examples (e.g. local functions) where we get `Undefined` for register rule when we cannot apply this logic. Example 1: local functions in main.rs will have LR rule as `Undefined`. Example 2: main()-> ! that is called from a trampoline will have a valid LR rule.
+    /// The iterator will continue until we have an LR register value of `None`. This will be true under the following conditions:
+    /// - We encounter a LR register value of 0xFFFFFFFF which is the 'Reset` value for that register.
+    /// - TODO: Catch the situation where the PC value indicates a hard-fault or other non-recoverable exception
+    /// - We can not intelligently infer a valid LR register value from the other registers.
+    /// - We legitimately get an LR register value of 0x0, indicating the bottom of the stack.
+    /// - Similarly, certain error conditions encountered in `StackFrameIterator` will also clear the LR register, to prevent further (likely faulty) iterations.
+    /// - If the FP register is 0x0, then we have encountered a logic 'bottom of the stack' (i.e. a non-returning function)
+    /// Note: In addition to populating the `StackFrame`s, this function will also populate the `DebugInfo::VariableCache` with `Variable`s for available Registers and function variables.
     fn next(&mut self) -> Option<Self::Item> {
-        let pc = match self.pc {
-            Some(pc) => pc,
-            None => {
-                log::debug!("Unable to determine next frame, program counter is unknown");
-                return None;
-            }
+        // PART 0-a: If we've encountered an error in the previous iteration, the `PC` will be `None`.
+        let frame_pc = if let Some(frame_pc) = self.unwind_registers.get_program_counter() {
+            frame_pc as u64
+        } else {
+            log::debug!(
+                "UNWIND: No available PC (program counter). Cannot continue stack unwinding."
+            );
+            return None;
         };
-
-        use gimli::UnwindSection;
-        let mut ctx = Box::new(gimli::UnwindContext::new());
-        let bases = gimli::BaseAddresses::default();
-
-        log::debug!("StackFrame: Unwinding at address {:#010x}", pc);
-
-        // Find function information, to check if we are in an inlined function.
-
-        let inline_call_site = match self.inlining_state {
-            InlineFunctionState::InlinedCallSite { .. } => true,
-            InlineFunctionState::NoInlining => false,
+        // PART 0-b: We cannot go deeper in the stack.
+        let frame_fp = if let Some(frame_fp) = self.unwind_registers.get_frame_pointer() {
```

Comment:
The frame pointer is not necessary for unwinding, since we use the debug info. So this check here is a bit too strict.

Reply:
@Tiwalun , I fixed this in the latest push. Please confirm that you are comfortable with the test results and 'Resolve conversation'